### ISSUE TITLE
conf-blas & conf-lapack fail to install under FreeBSD

### ISSUE THREAD

**Issue by magv** [ocaml/opam-repository#7330]
When installing `conf-blas` under FreeBSD 10 amd64 I get this error:

```
$ opam install conf-blas                                                                                            
The following actions will be performed:
  ∗  install conf-blas 1

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of conf-blas failed at "sh -exc cc $CFLAGS test.c -lblas".

#=== ERROR while installing conf-blas.1 =======================================#
# opam-version 1.2.2
# os           freebsd
# command      sh -exc cc $CFLAGS test.c -lblas
# path         ~/.opam/4.03.0+flambda/build/conf-blas.1
# compiler     4.03.0+flambda
# exit-code    1
# env-file     ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32329-ccabea.env
# stdout-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32329-ccabea.out
# stderr-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32329-ccabea.err
### stderr ###
# + cc test.c -lblas
# /usr/bin/ld: cannot find -lblas
# cc: error: linker command failed with exit code 1 (use -v to see invocation)



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions failed
  ∗  install conf-blas 1
No changes have been performed

=-=- conf-blas.1 troobleshooting =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
=> This package relies on external (system) dependencies that may be missing. `opam depext conf-blas.1' may help you find the correct
   installation for your system.
```

Adding `-L/usr/local/lib` to `CFLAGS`, so the build script would find `libblas.so`, results in a new set of errors:

```
#=== ERROR while installing conf-blas.1 =======================================#
# opam-version 1.2.2
# os           freebsd
# command      sh -exc cc $CFLAGS test.c -lblas
# path         /home/magv/.opam/4.03.0+flambda/build/conf-blas.1
# compiler     4.03.0+flambda
# exit-code    1
# env-file     ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32343-ccabea.env
# stdout-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32343-ccabea.out
# stderr-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32343-ccabea.err
### stderr ###
# [...]
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__addtf3@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__gttf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__divtf3@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__letf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__netf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__floatditf@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__trunctfdf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__eqtf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__floatsitf@GCC_4.6.0'
# cc: error: linker command failed with exit code 1 (use -v to see invocation)
```

This is probably because BLAS under FreeBSD is build with GCC, while the default `cc` is CLANG, and there seem to be incompatibilities between them. Whatever the case, here's a change to `conf-blas` that seems to make it work for me:

``` diff
--- ~/.opam/repo/default/packages/conf-blas/conf-blas.1/opam.orig      2016-08-31 15:26:25.000000000 +0300
+++ ~/.opam/repo/default/packages/conf-blas/conf-blas.1/opam   2016-08-31 15:42:09.000000000 +0300
@@ -6,7 +6,8 @@
 dev-repo: "https://github.com/ocaml/opam-repository.git"
 license: "GPL-2.0"
 build: [
-  ["sh" "-exc" "cc $CFLAGS test.c -lblas"] {os != "darwin" & os != "win32"}
+  ["sh" "-exc" "cc $CFLAGS test.c -lblas"] {os != "darwin" & os != "freebsd" & os != "win32"}
+  ["sh" "-exc" "${CC:-gcc} $CFLAGS test.c -lblas $LDFLAGS"] {os = "freebsd"}
   ["sh" "-exc" "cc -framework Accelerate $CFLAGS test.c -lblas"] {os = "darwin"}
   ["%{build}%/test-win.sh"] {os = "win32"}
 ]
@@ -15,4 +16,5 @@
   [["ubuntu"] ["libblas-dev"]]
   [["centos"] ["blas-devel"]]
   [["fedora"] ["blas-devel"]]
+  [["freebsd"] ["blas" "gcc"]]
 ]
```

(I made it possible to override the compiler with `CC` variable to make it easier to debug the thing if something will break in the future).

Here's a similar change for `conf-lapack`, which has the same set of problems:

``` diff
--- ~/.opam/repo/default/packages/conf-lapack/conf-lapack.1/opam.orig  2016-08-31 16:13:36.000000000 +0300
+++ ~/.opam/repo/default/packages/conf-lapack/conf-lapack.1/opam       2016-08-31 16:13:40.000000000 +0300
@@ -6,7 +6,8 @@
 dev-repo: "https://github.com/ocaml/opam-repository.git"
 license: "GPL-2.0"
 build: [
-  ["sh" "-exc" "cc $CFLAGS test.c -llapack"] {os != "darwin" & os != "win32"}
+  ["sh" "-exc" "cc $CFLAGS test.c -llapack"] {os != "darwin" & os != "freebsd" & os != "win32"}
+  ["sh" "-exc" "${CC:-gcc} $CFLAGS test.c -llapack $LDFLAGS"] {os = "freebsd"}
   ["sh" "-exc" "cc -framework Accelerate $CFLAGS test.c -llapack"] {os = "darwin"}
   ["%{build}%/test-win.sh"] {os = "win32"}
 ]
@@ -15,4 +16,5 @@
   [["ubuntu"] ["liblapack-dev"]]
   [["centos"] ["lapack-devel"]]
   [["fedora"] ["lapack-devel"]]
+  [["freebsd"] ["lapack" "gcc"]]
 ]
```

With both of these changes I'm able to install `lacaml`, and it seems to work too.

(Sorry for providing patches inline without a pull request, but that is such a hassle).

**Comment by mmottl** [comment]
@magv Thanks for the patch, I have submitted a pull request addressing this issue.

**Comment by dbuenzli** [comment]
This above PR was merged manually by @yallop in 57954d5f43aad222d73fa70 and ccfdcaa5c6544e21ea0cbc70f3e1a6c356 and this should be solved.
